File: lib/nfl/auction.ts
/**
 * AuctionEngine — the server-authoritative core of the NFL Auction game.
 *
 * State is a plain serializable object (persisted to Redis). All mutations go
 * through pure reducers that VALIDATE before applying, so the client can never
 * force an illegal budget, duplicate player, over-bid, or impossible roster.
 */

import {
  type AuctionLot,
  type AuctionResult,
  type BidRecord,
  type NflGameConfig,
  type NflGameResult,
  type GameStatus,
  type RosterState,
  type Season,
  type NflPlayer,
  type TeamId,
} from "./types"
import { getSeasonPlayers } from "./data"
import { emptyRoster, isRosterComplete, canAddPlayer, ownsPlayer, placePlayer, bestSlotFor } from "./roster"
import { maxAffordable, remaining, validateBidAmount, MIN_BID } from "./budget"
import { scaledOpeningBid } from "./value"
import { mulberry32, hashSeed, shuffle, type RNG } from "./rng"

// ─── Serializable game state ─────────────────────────────────────────────────

export interface NflAuctionState {
  gameId: string
  seed: number
  season: Season
  config: NflGameConfig
  status: GameStatus
  /** Ordered player ids still to be auctioned (front = current lot). */
  queue: string[]
  lot: AuctionLot | null
  passed: TeamId[]
  rosters: Record<TeamId, RosterState>
  isAI: Record<TeamId, boolean>
  history: BidRecord[]
  results: AuctionResult[]
  lotDeadline: number | null
  completedAt: number | null
  result?: NflGameResult | null
}

export function otherTeam(t: TeamId): TeamId {
  return t === "P1" ? "P2" : "P1"
}

/**
 * Total lots put up for auction. Both players need 9 (18 total); we show a few
 * extra so there's genuine competition and choice, but cap it so the auction
 * doesn't drag through the entire pool.
 */
export const MAX_LOTS = 24

/**
 * Build the auction order. For small budgets (≤$25) the biggest stars come
 * LAST so the "can you still afford a stud?" drama survives the cap. Larger
 * budgets interleave tiers for a lively, unpredictable flow. Shuffled within
 * tiers for variety.
 */
export function buildAuctionOrder(rng: RNG, pool: NflPlayer[], budget = 25): string[] {
  const byTier: Record<number, NflPlayer[]> = { 1: [], 2: [], 3: [], 4: [] }
  for (const p of pool) byTier[p.tier].push(p)
  for (const t of [1, 2, 3, 4]) byTier[t] = shuffle(rng, byTier[t])

  const cap = Math.min(MAX_LOTS, pool.length)

  if (budget <= 25) {
    const starCount = Math.min(6, byTier[1].length)
    const stars = byTier[1].slice(0, starCount)
    const restQuota = cap - stars.length
    const front = [...byTier[4], ...byTier[3], ...byTier[2]].slice(0, restQuota)
    return [...front, ...stars].map((p) => p.id)
  }

  const order: NflPlayer[] = []
  const buckets = [byTier[3], byTier[4], byTier[2], byTier[1]]
  let idx = 0
  while (buckets.some((b) => b.length > 0) && order.length < cap) {
    const bucket = buckets[idx % buckets.length]
    if (bucket.length > 0) order.push(bucket.shift()!)
    idx++
    if (idx % 4 === 0 && rng() < 0.4) idx++
  }
  return order.slice(0, cap).map((p) => p.id)
}

export function createGame(opts: {
  gameId: string
  config: NflGameConfig
  seed?: number
  vsAI?: boolean
}): NflAuctionState {
  const seed = opts.seed ?? hashSeed(opts.gameId)
  const rng = mulberry32(seed)
  const pool = getSeasonPlayers(opts.config.season)
  const queue = buildAuctionOrder(rng, pool, opts.config.budgetPerPlayer)

  return {
    gameId: opts.gameId,
    seed,
    season: opts.config.season,
    config: opts.config,
    status: "auction",
    queue,
    lot: null,
    passed: [],
    rosters: { P1: emptyRoster(), P2: emptyRoster() },
    isAI: { P1: false, P2: opts.vsAI !== false },
    history: [],
    results: [],
    lotDeadline: null,
    completedAt: null,
  }
}

// ─── Lot lifecycle ─────────────────────────────────────────────────────────

function playerById(state: NflAuctionState, id: string): NflPlayer {
  const p = getSeasonPlayers(state.season).find((x) => x.id === id)
  if (!p) throw new Error(`Unknown player id: ${id}`)
  return p
}

/** Skip queued players NEITHER team can roster anymore. Guarantees termination. */
function pruneUnwinnableFront(state: NflAuctionState): void {
  while (state.queue.length > 0) {
    const p = playerById(state, state.queue[0])
    const anyCanTake =
      (!isRosterComplete(state.rosters.P1) && canAddPlayer(state.rosters.P1, p)) ||
      (!isRosterComplete(state.rosters.P2) && canAddPlayer(state.rosters.P2, p))
    if (anyCanTake) break
    state.queue.shift()
  }
}

/** Open the next lot. Returns true if a lot was opened, false if none remain. */
export function openNextLot(state: NflAuctionState): boolean {
  if (rostersDone(state)) {
    finalizeAuction(state)
    return false
  }
  pruneUnwinnableFront(state)
  if (state.queue.length === 0) {
    finalizeAuction(state)
    return false
  }
  const id = state.queue[0]
  const player = playerById(state, id)
  let open = scaledOpeningBid(player, state.config.budgetPerPlayer, state.config.rosterSize)

  // Clamp the opening bid to what an active, eligible team can actually afford,
  // so a lot can never be un-sellable and stall the auction.
  let affordableFloor = 0
  for (const t of ["P1", "P2"] as TeamId[]) {
    if (isRosterComplete(state.rosters[t])) continue
    if (!canAddPlayer(state.rosters[t], player)) continue
    affordableFloor = Math.max(affordableFloor, maxAffordable(state.config.budgetPerPlayer, state.rosters[t]))
  }
  if (affordableFloor > 0) open = Math.min(open, affordableFloor)
  open = Math.max(1, open)

  state.lot = { player, currentBid: open, highBidder: null, openingBid: open }
  state.passed = []
  state.lotDeadline = Date.now() + state.config.auctionTimerSeconds * 1000
  return true
}

function rostersDone(state: NflAuctionState): boolean {
  return isRosterComplete(state.rosters.P1) && isRosterComplete(state.rosters.P2)
}

/** A team is out of the auction once its roster is full. */
export function teamActive(state: NflAuctionState, team: TeamId): boolean {
  return !isRosterComplete(state.rosters[team])
}

// ─── Bidding ─────────────────────────────────────────────────────────────

export interface BidOutcome {
  ok: boolean
  error?: string
  state: NflAuctionState
}

/**
 * Place a bid for `team` at `amount`. Fully validated & server-authoritative.
 * On success the team becomes high bidder, passes are cleared, and the lot timer
 * resets to the shorter window.
 */
export function placeBid(state: NflAuctionState, team: TeamId, amount: number): BidOutcome {
  if (state.status !== "auction" || !state.lot) {
    return { ok: false, error: "No active lot.", state }
  }
  if (!teamActive(state, team)) {
    return { ok: false, error: "Your roster is already full.", state }
  }
  const player = state.lot.player
  if (ownsPlayer(state.rosters[team], player.id)) {
    return { ok: false, error: "You already own this player.", state }
  }
  if (!canAddPlayer(state.rosters[team], player)) {
    return { ok: false, error: `You have no open slot for ${player.name}.`, state }
  }
  const total = state.config.budgetPerPlayer
  const allowEqual = state.lot.highBidder === null
  const err = validateBidAmount(amount, state.lot.currentBid, total, state.rosters[team], allowEqual)
  if (err) return { ok: false, error: err, state }

  state.lot.currentBid = amount
  state.lot.highBidder = team
  state.passed = [] // any new bid re-opens the lot for both
  state.history.push({
    playerId: player.id,
    playerName: player.name,
    bidder: team,
    amount,
    ts: Date.now(),
  })
  state.lotDeadline = Date.now() + state.config.auctionTimerResetSeconds * 1000
  return { ok: true, state }
}

/**
 * Next legal bid for a team = currentBid + league increment, capped by
 * affordability. Returns null if they can't afford even one step.
 */
export function minRaise(state: NflAuctionState, team: TeamId): number | null {
  if (!state.lot) return null
  const max = maxAffordable(state.config.budgetPerPlayer, state.rosters[team])

  if (state.lot.highBidder === null) {
    const open = state.lot.currentBid
    return open <= max ? open : null
  }

  const step = state.config.bidIncrement || 1
  const next = state.lot.currentBid + step
  if (next > max) {
    const one = state.lot.currentBid + 1
    return one <= max ? one : null
  }
  return next
}

/** A team passes on the current lot; may resolve the lot. */
export function pass(state: NflAuctionState, team: TeamId): BidOutcome {
  if (state.status !== "auction" || !state.lot) {
    return { ok: false, error: "No active lot.", state }
  }
  if (!state.passed.includes(team)) state.passed.push(team)
  maybeResolveLot(state)
  return { ok: true, state }
}

function biddableTeams(state: NflAuctionState): TeamId[] {
  const teams: TeamId[] = ["P1", "P2"]
  return teams.filter((t) => canBidOnLot(state, t))
}

function canBidOnLot(state: NflAuctionState, team: TeamId): boolean {
  if (!state.lot) return false
  if (!teamActive(state, team)) return false
  if (ownsPlayer(state.rosters[team], state.lot.player.id)) return false
  if (!canAddPlayer(state.rosters[team], state.lot.player)) return false
  if (state.lot.highBidder === team) return true
  const next = state.lot.currentBid + 1
  return next <= maxAffordable(state.config.budgetPerPlayer, state.rosters[team])
}

/** Resolve the lot if all non-leaders who could bid have passed. */
export function maybeResolveLot(state: NflAuctionState): void {
  if (!state.lot) return
  const eligible = biddableTeams(state).filter((t) => t !== state.lot!.highBidder)
  const allNonLeadersPassed = eligible.every((t) => state.passed.includes(t))
  if (allNonLeadersPassed) resolveLot(state)
}

/** Force-resolve the current lot (timer expiry / all passed). */
export function resolveLot(state: NflAuctionState): void {
  if (!state.lot) return
  const { player, highBidder, currentBid } = state.lot

  if (highBidder && teamActive(state, highBidder)) {
    const roster = state.rosters[highBidder]
    const slot = bestSlotFor(roster, player)
    if (slot) {
      state.rosters[highBidder] = placePlayer(roster, player, currentBid, slot)
      state.results.push({ playerId: player.id, playerName: player.name, winner: highBidder, price: currentBid })
    }
  }
  state.queue = state.queue.filter((id) => id !== player.id)
  state.lot = null
  state.passed = []
  state.lotDeadline = null
  openNextLot(state)
}

/** Server tick: resolve an expired lot. Returns true if state changed. */
export function tick(state: NflAuctionState, now: number = Date.now()): boolean {
  if (state.status !== "auction") return false
  if (!state.lot) return openNextLot(state)
  if (state.lotDeadline !== null && now >= state.lotDeadline) {
    resolveLot(state)
    return true
  }
  return false
}

// ─── Finalization ──────────────────────────────────────────────────────────

/**
 * Auto-fill any remaining slots with the best affordable free agents so the
 * game always reaches a valid, playable state, then flip to "lineup".
 */
export function finalizeAuction(state: NflAuctionState): void {
  autoFillIfNeeded(state, "P1")
  autoFillIfNeeded(state, "P2")
  state.lot = null
  state.passed = []
  state.lotDeadline = null
  state.status = "lineup"
}

function autoFillIfNeeded(state: NflAuctionState, team: TeamId): void {
  if (isRosterComplete(state.rosters[team])) return
  const pool = getSeasonPlayers(state.season)
  const owned = new Set<string>()
  for (const t of ["P1", "P2"] as TeamId[]) {
    for (const s of Object.values(state.rosters[t].slots)) {
      if (s) owned.add(s.player.id)
    }
  }
  const available = pool.filter((p) => !owned.has(p.id)).sort((a, b) => b.overall - a.overall)

  let guard = 0
  while (!isRosterComplete(state.rosters[team]) && guard++ < 60) {
    let placed = false
    for (const p of available) {
      if (owned.has(p.id)) continue
      if (!canAddPlayer(state.rosters[team], p)) continue
      const rem = remaining(state.config.budgetPerPlayer, state.rosters[team])
      const price = Math.min(MIN_BID, Math.max(0, rem))
      state.rosters[team] = placePlayer(state.rosters[team], p, price)
      owned.add(p.id)
      state.results.push({ playerId: p.id, playerName: p.name, winner: team, price })
      placed = true
      break
    }
    if (!placed) break
  }
}

File: auction_ai/simulation/auction.py
"""Auction lifecycle — faithful port of lib/arena/auction.ts.

Reproduces order building, lot selection, bidding, resolution, re-offers, and
end-of-auction auto-fill so the Python simulator plays by the EXACT rules the
production engine (and therefore real users) will experience.

The one intentional difference from the server: there is no wall clock. A CPU
that declines to raise "stands pat" (the same behavior as the live server's
driveAI), and the lot resolves once every biddable non-leader has effectively
declined. The simulator's drive loop (auction_simulator.py) reproduces the
"nobody will raise → resolve" settlement.
"""
from __future__ import annotations

from typing import Callable, Dict, List, Optional

from .auction_state import (
    ArenaState,
    AuctionLot,
    AuctionResultRow,
    BidRecord,
    GameConfig,
    TEAMS,
)
from .budget import max_affordable, remaining, validate_bid_amount
from .player_model import (
    MAX_LOTS,
    MAX_REOFFERS,
    MIN_BID,
    MIN_PER_POSITION,
    POSITIONS,
    Player,
    get_season_players,
    players_by_id,
)
from .rng import hash_seed, mulberry32, shuffle
from .roster import (
    auction_slot_for,
    can_add_player,
    can_force_add_player,
    empty_roster,
    is_roster_complete,
    open_starter_slots,
    owns_player,
    place_player,
)
from .value import scaled_opening_bid


# ─── Order building ─────────────────────────────────────────────────────────
def _ensure_positional_coverage(
    selected: List[Player], pool: List[Player]
) -> List[Player]:
    out = list(selected)
    in_board = {p.id for p in out}
    leftovers = [p for p in pool if p.id not in in_board]

    def count_for(pos: str, lst: List[Player]) -> int:
        return sum(1 for p in lst if pos in p.eligible_positions())

    for pos in POSITIONS:
        while count_for(pos, out) < MIN_PER_POSITION:
            cand_idx = next(
                (i for i, p in enumerate(leftovers) if pos in p.eligible_positions()),
                -1,
            )
            if cand_idx == -1:
                break
            drop_idx = -1
            for i in range(len(out) - 1, -1, -1):
                p = out[i]
                if pos in p.eligible_positions():
                    continue
                still_covered = all(
                    count_for(pp, out) > MIN_PER_POSITION
                    for pp in p.eligible_positions()
                )
                if still_covered:
                    drop_idx = i
                    break
            candidate = leftovers.pop(cand_idx)
            if drop_idx >= 0:
                out[drop_idx] = candidate
            else:
                out.append(candidate)
    return out


def build_auction_order(rng: Callable[[], float], pool: List[Player], budget: int = 25) -> List[str]:
    by_tier: Dict[int, List[Player]] = {1: [], 2: [], 3: [], 4: []}
    for p in pool:
        by_tier[p.tier].append(p)
    for t in (1, 2, 3, 4):
        by_tier[t] = shuffle(rng, by_tier[t])

    cap = min(MAX_LOTS, len(pool))

    if budget <= 25:
        star_count = min(4, len(by_tier[1]))
        stars = by_tier[1][:star_count]
        rest_quota = cap - len(stars)
        front = shuffle(rng, [*by_tier[4], *by_tier[3], *by_tier[2]])[:rest_quota]
        covered = _ensure_positional_coverage(
            front, [*by_tier[4], *by_tier[3], *by_tier[2]]
        )
        return [p.id for p in [*covered, *stars]]

    order: List[Player] = []
    buckets = [by_tier[3], by_tier[4], by_tier[2], by_tier[1]]
    idx = 0
    while any(len(b) > 0 for b in buckets) and len(order) < cap:
        bucket = buckets[idx % len(buckets)]
        if bucket:
            order.append(bucket.pop(0))
        idx += 1
        if idx % 4 == 0 and rng() < 0.4:
            idx += 1
    return [p.id for p in _ensure_positional_coverage(order[:cap], pool)]


def create_game(
    game_id: str,
    config: GameConfig,
    seed: Optional[int] = None,
) -> ArenaState:
    s = seed if seed is not None else hash_seed(game_id)
    rng = mulberry32(s)
    pool = get_season_players(config.season)
    queue = build_auction_order(rng, pool, config.budget_per_player)
    return ArenaState(
        game_id=game_id,
        seed=s,
        season=config.season,
        config=config,
        queue=queue,
        rosters={"P1": empty_roster(), "P2": empty_roster()},
    )


# ─── Lot lifecycle ──────────────────────────────────────────────────────────
def _player_by_id(state: ArenaState, pid: str) -> Player:
    p = players_by_id(state.season).get(pid)
    if p is None:
        raise ValueError(f"Unknown player id: {pid}")
    return p


def rosters_done(state: ArenaState) -> bool:
    return is_roster_complete(state.rosters["P1"]) and is_roster_complete(
        state.rosters["P2"]
    )


def team_active(state: ArenaState, team: str) -> bool:
    return not is_roster_complete(state.rosters[team])


def active_teams(state: ArenaState) -> List[str]:
    return [t for t in TEAMS if not is_roster_complete(state.rosters[t])]


def _prune_unrosterable(state: ArenaState) -> None:
    def keep(pid: str) -> bool:
        p = _player_by_id(state, pid)
        return (
            not is_roster_complete(state.rosters["P1"])
            and can_force_add_player(state.rosters["P1"], p)
        ) or (
            not is_roster_complete(state.rosters["P2"])
            and can_force_add_player(state.rosters["P2"], p)
        )

    state.queue = [pid for pid in state.queue if keep(pid)]


def _relaxed_opening_floor(state: ArenaState, player: Player, team: str) -> Optional[int]:
    if player.tier <= 2:
        return None
    roster = state.rosters[team]
    if is_roster_complete(roster):
        return None
    if not can_add_player(roster, player):
        return None
    reserve = scaled_opening_bid(
        player, state.config.budget_per_player, state.config.roster_size
    )
    mx = max_affordable(state.config.budget_per_player, roster)
    if mx >= reserve:
        return None
    return mx if mx >= MIN_BID else None


def effective_opening_bid(state: ArenaState, player: Player) -> int:
    reserve = scaled_opening_bid(
        player, state.config.budget_per_player, state.config.roster_size
    )
    floor = reserve
    for t in TEAMS:
        clamp = _relaxed_opening_floor(state, player, t)
        if clamp is not None:
            floor = min(floor, max(MIN_BID, min(reserve, clamp)))
    return max(MIN_BID, floor)


def _select_next_lot(state: ArenaState) -> bool:
    teams = active_teams(state)
    if not teams:
        return False
    queued = [_player_by_id(state, pid) for pid in state.queue]

    def biddable_filter(p: Player) -> bool:
        opening = effective_opening_bid(state, p)
        return any(
            can_add_player(state.rosters[t], p)
            and max_affordable(state.config.budget_per_player, state.rosters[t]) >= opening
            for t in teams
        )

    biddable = [p for p in queued if biddable_filter(p)]
    if not biddable:
        return False

    demand = {
        pos: sum(1 for t in teams if state.rosters[t].slots[pos] is None)
        for pos in POSITIONS
    }
    supply = {
        pos: sum(1 for p in queued if pos in p.eligible_positions())
        for pos in POSITIONS
    }
    critical = {pos for pos in POSITIONS if demand[pos] > 0 and supply[pos] <= demand[pos]}

    def priority_of(p: Player) -> int:
        best = 0
        for t in teams:
            if not can_add_player(state.rosters[t], p):
                continue
            open_ = open_starter_slots(state.rosters[t])
            fills = [pos for pos in p.eligible_positions() if pos in open_]
            if not fills:
                continue
            best = max(best, 2 if any(pos in critical for pos in fills) else 1)
        return best

    best = biddable[0]
    best_score = -1
    for p in biddable:
        score = priority_of(p)
        if score > best_score:
            best_score = score
            best = p
            if score == 2:
                break

    idx = state.queue.index(best.id)
    if idx > 0:
        chosen = state.queue.pop(idx)
        state.queue.insert(0, chosen)
    return True


def open_next_lot(state: ArenaState) -> bool:
    if rosters_done(state):
        finalize_auction(state)
        return False
    _prune_unrosterable(state)
    if not state.queue:
        finalize_auction(state)
        return False
    if not _select_next_lot(state):
        finalize_auction(state)
        return False
    pid = state.queue[0]
    player = _player_by_id(state, pid)
    opening = effective_opening_bid(state, player)
    state.lot = AuctionLot(
        player_id=pid, current_bid=opening, high_bidder=None, opening_bid=opening
    )
    state.passed = []
    return True


# ─── Bidding ────────────────────────────────────────────────────────────────
class BidOutcome:
    def __init__(self, ok: bool, error: Optional[str] = None):
        self.ok = ok
        self.error = error


def place_bid(state: ArenaState, team: str, amount: int) -> BidOutcome:
    if state.status != "auction" or state.lot is None:
        return BidOutcome(False, "No active lot.")
    if not team_active(state, team):
        return BidOutcome(False, "Your roster is already full.")
    if state.lot.high_bidder == team:
        return BidOutcome(False, "You're already the high bidder.")
    player = _player_by_id(state, state.lot.player_id)
    if owns_player(state.rosters[team], player.id):
        return BidOutcome(False, "You already own this player.")
    if not can_add_player(state.rosters[team], player):
        return BidOutcome(False, f"No open slot for {player.name}.")
    total = state.config.budget_per_player
    allow_equal = state.lot.high_bidder is None
    err = validate_bid_amount(
        amount, state.lot.current_bid, total, state.rosters[team], allow_equal
    )
    if err:
        return BidOutcome(False, err)

    state.lot.current_bid = amount
    state.lot.high_bidder = team
    state.passed = []
    state.history.append(BidRecord(player_id=player.id, bidder=team, amount=amount))
    return BidOutcome(True)


def min_raise(state: ArenaState, team: str) -> Optional[int]:
    if state.lot is None:
        return None
    mx = max_affordable(state.config.budget_per_player, state.rosters[team])
    if state.lot.high_bidder is None:
        opening = state.lot.current_bid
        return opening if opening <= mx else None
    step = state.config.bid_increment or 1
    nxt = state.lot.current_bid + step
    if nxt > mx:
        one = state.lot.current_bid + 1
        return one if one <= mx else None
    return nxt


def passes(state: ArenaState, team: str) -> BidOutcome:
    if state.status != "auction" or state.lot is None:
        return BidOutcome(False, "No active lot.")
    if team not in state.passed:
        state.passed.append(team)
    maybe_resolve_lot(state)
    return BidOutcome(True)


def _can_bid_on_lot(state: ArenaState, team: str) -> bool:
    if state.lot is None:
        return False
    if not team_active(state, team):
        return False
    player = _player_by_id(state, state.lot.player_id)
    if owns_player(state.rosters[team], player.id):
        return False
    if not can_add_player(state.rosters[team], player):
        return False
    if state.lot.high_bidder == team:
        return True
    nxt = state.lot.current_bid + 1
    return nxt <= max_affordable(state.config.budget_per_player, state.rosters[team])


def biddable_teams(state: ArenaState) -> List[str]:
    return [t for t in TEAMS if _can_bid_on_lot(state, t)]


def maybe_resolve_lot(state: ArenaState) -> None:
    if state.lot is None:
        return
    eligible = [t for t in biddable_teams(state) if t != state.lot.high_bidder]
    if all(t in state.passed for t in eligible):
        resolve_lot(state)


def resolve_lot(state: ArenaState) -> None:
    if state.lot is None:
        return
    lot = state.lot
    player = _player_by_id(state, lot.player_id)
    sold = False
    if lot.high_bidder and team_active(state, lot.high_bidder):
        roster = state.rosters[lot.high_bidder]
        slot = auction_slot_for(roster, player)
        if slot:
            state.rosters[lot.high_bidder] = place_player(
                roster, player, lot.current_bid, slot
            )
            state.results.append(
                AuctionResultRow(
                    player_id=player.id, winner=lot.high_bidder, price=lot.current_bid
                )
            )
            sold = True

    state.queue = [pid for pid in state.queue if pid != player.id]

    if not sold:
        seen = state.reoffers.get(player.id, 0)
        anyone_needs = not is_roster_complete(state.rosters["P1"]) or not is_roster_complete(
            state.rosters["P2"]
        )
        if anyone_needs and seen < MAX_REOFFERS:
            state.reoffers[player.id] = seen + 1
            state.queue.append(player.id)

    state.lot = None
    state.passed = []
    open_next_lot(state)


# ─── Finalization ─────────────────────────────────────────────────────────
def finalize_auction(state: ArenaState) -> None:
    _auto_fill_if_needed(state, "P1")
    _auto_fill_if_needed(state, "P2")
    state.lot = None
    state.passed = []
    state.status = "lineup"


def _auto_fill_if_needed(state: ArenaState, team: str) -> None:
    if is_roster_complete(state.rosters[team]):
        return
    pool = get_season_players(state.season)
    owned = set()
    for t in TEAMS:
        for o in state.rosters[t].slots.values():
            if o:
                owned.add(o.player.id)
    available = sorted(
        (p for p in pool if p.id not in owned),
        key=lambda p: (p.overall, p.starting_bid),
    )
    guard = 0
    while not is_roster_complete(state.rosters[team]) and guard < 50:
        guard += 1
        placed = False
        for p in available:
            if p.id in owned:
                continue
            if not can_force_add_player(state.rosters[team], p):
                continue
            rem = remaining(state.config.budget_per_player, state.rosters[team])
            price = min(MIN_BID, max(0, rem))
            state.rosters[team] = place_player(state.rosters[team], p, price)
            owned.add(p.id)
            state.results.append(
                AuctionResultRow(player_id=p.id, winner=team, price=price)
            )
            placed = True
            break
        if not placed:
            break
